src/coreclr/gc/gc.cpp

```diff
@@ -9349,6 +9367,60 @@ size_t gc_heap::sort_mark_list()
     }
 #endif //USE_VXSORT
 
+    uint8_t** x = mark_list;
+
+#ifdef USE_REGIONS
+    // first set the pieces for all regions to empty
+    size_t lowest_region_index = (size_t)g_gc_lowest_address >> gc_heap::min_segment_size_shr;
+    size_t highest_region_index = (size_t)g_gc_highest_address >> gc_heap::min_segment_size_shr;
+    size_t region_count = highest_region_index - lowest_region_index + 1;
+    if (mark_list_piece_size < region_count)
+    {
+        delete[] mark_list_piece_start;
+        delete[] mark_list_piece_end;
+
+        mark_list_piece_start = new (nothrow) uint8_t** [region_count];
+        mark_list_piece_end   = new (nothrow) uint8_t** [region_count];
+        if (mark_list_piece_start && mark_list_piece_end)
+        {
+            mark_list_piece_size = region_count;
+        }
+        else
+        {
+            if (mark_list_piece_start)
+                delete[] mark_list_piece_start;
+            if (mark_list_piece_end)
+                delete[] mark_list_piece_end;
+            mark_list_piece_size = 0;
+            return 0;
```

Comment:
we'd want to fake a mark list overflow here as well before returning.

Reply:
Right, the failure case handling is incomplete. I think faking a mark list overflow doesn't quite do it, rather I'll need to check mark_list_piece_size for *all* the heaps in plan_phase when we compute use_mark_list.

Reply:
what I meant was to fake a mark list overflow like you did with this code toward the beginning of sort_mark_list so you didn't need to have plan phase check for the mark_list_piece_size 

```cplusplus
    // if any other heap had a mark list overflow, we fake one too,
    // so we don't use an incomplete mark list by mistake
```
however that's also not correct 'cause some heaps are already using their mark_list_index so it'd be wrong to change it. but now I've thought about it a bit more I think we could make this better by allocating the memory for the new mark_list_piece*, if needed, before sort_mark_list. so for regions we'd have something like grow_mark_list_pieces before the `gc_join_null_dead_long_weak` join. and in the join we fake a global mark list overflow if any heap failed to alloc memory for their mark_list_piece*. this way we don't need to call sort_mark_list at all on any heap if we already don't have memory for the pieces (so we save the sort - of course you can achieve that part by simply moving the alloc code toward the top in sort_mark_list for that heap; but this way we save it for all heaps). what do you think? am I missing anything?

another thing we can do to optimize is to get the actual region count but that can be part of the server GC work.

Reply:
Well, perhaps it would be best to do the allocation at the very start of the mark phase and disable the mark list if we don't succeed, similarly to how we disable it for gen 2 GCs?

Reply:
that's fine too...doesn't make a big difference to me.